### ISSUE TITLE
Metrics: Active/current operations are not counted in duration_in_millis

### ISSUE THREAD

**Issue by jordansissel** [elastic/logstash#5683]
Based on discussion in https://github.com/widhalmt/check_logstash/issues/6, it's probably useful to be able to see any active operations via metrics to see how long any timed operations have been running. 

The problem is that the way we accumulate time for `duration` metrics is by counting the duration of each operation and adding that duration to the metric. However, because we only add _after_ the operation completes, any in-flight operations are not visible in the duration metrics.

---

I have two proposals for how we could share this info:
- Any active timers are included in the `duration_in_millis` (or similar) metrics when queried. This will allow us to see growth caused by incomplete-but-long-running operations
- or, show all active timers by metric name as well as their start times and current duration.

**Comment by ph** [comment]
We have discussed on zoom And I think the changes are pretty minimal.

All the time are calculated inside the `TimedException` class and it uses a gauge behind the scene as the data holder.

So we could do the following, when we create a new `TimedException` we record the start time, we can wrap the start time into a special object that will be saved inside the the gauge and that object can implement a `to_json` or `to_json_data` that the api will correctly take when it tries to serialize the data for the web api.


```ruby
class TimedExecutionStarted
  def initialize
    @start = Time.now
  end

 def duration
  Time.now - @start
 end
  
  # options A
  def to_json
    "Timer started #{@start - Time.now}s ago." # uses I18N.
  end

# options B
  def to_json
    { :status => "incomplete", started_at => @start, :duration => ago }
  end
end
```

@jordansissel What do you think about theses two options A-B? B is probably more machine friendly.

**Comment by ph** [comment]
I am wondering if this wont be confusing with option B, when the timer complete the hash will be replacer by a int.

**Comment by jordansissel** [comment]
@ph yeah, I was thinking we'd have to present active TimedExecutions somewhere else in the metric namespace, especially in cases where you have parallel TimedExecutions with the same metric name.

I like option B, but presenting that information somewhere other than name used by the timer itself -- does that make sense?

**Comment by jordansissel** [comment]
Like, if we have:

```
clock = metric.time(:duration_in_millis)
...
clock.stop
```

This would produce a metric space similar to:

```
{ "duration_in_millis": 12345 }
```

Maybe we can have something like 'active execution' as a sibling namespace:

```
{
   "duration_in_millis": 12345,
   "active_execution": {
      "what goes here?": { "start_in_epoch_millis": 1469183723123, "duration_in_millis": 1414 },
      ...
   }
}
```

**Comment by jordansissel** [comment]
As a completely alternate possibility, maybe this is an opportunity for a new feature where we track transactions within a thread?

Have a per-thread logger that lets you log the current operation, and we can have a new API endpoint that shows all threads' current operation + start time + duration?

**Comment by ph** [comment]
The `active_execution` would mean that we need to clear them when their execution are actually complete. We don't have that option in the metric store yet to remove a previously recorded metric. 

I need to think about the implication of maintaining the 2 data structure that we have, I wonder how it will affect performance.


> As a completely alternate possibility, maybe this is an opportunity for a new feature where we track transactions within a thread? 

I like the concept, So the api would just display the current logger the threads?

**Comment by jordansissel** [comment]
I think we can defer or delay this feature for now. After talking with @acchen97 a bit, he pointed out that there are two troubleshooting scenarios to think about:

1) Troubleshooting periodic or past behavioral problems
2) Troubleshooting active behavioral problems

My concern was this: only updating `duration_in_millis` after an operation completes will cause `duration_in_millis` to not represent any in-flight operations, only completed ones. 

However, I think debugging active problems probably requires a different set of tools (Elasticsearch has `/_cat/pending_tasks` and `hot_threads` and similar, for example).

So I won't worry about solving this right now. We can work on it as time permits some time later.

**Comment by ph** [comment]
@jordansissel @acchen97 I think it make complete sense to have a different endpoints for theses kind of scenario. +1

**Comment by yaauie** [comment]
Resolved in 8.7.0 with live-timers via https://github.com/elastic/logstash/issues/14675
